proposals/1984_ca-bundle-from-secret-ref/README.md

```diff
@@ -11,124 +11,261 @@ Once approved and implemented, they become historical documents.
+
+## Implementation
+The custom resources have an optional `certificateAuthorityData` field to hold the CA bundle to be used for client-side TLS verification as part of the `TLSSpec` that is currently defined as follows.
+
+```go
+// Configuration for configuring TLS on various authenticators.
+type TLSSpec struct {
+	// X.509 Certificate Authority (base64-encoded PEM bundle). If omitted, a default set of system roots will be trusted.
+	// +optional
+	CertificateAuthorityData string `json:"certificateAuthorityData,omitempty"`
+}
+```
+
+To allow the custom resources to carry a reference to kubernetes secrets or configmaps,  the `TLSSpec` will be modified to look like
+```go
+// Reference to a kubernetes secrets and configmaps as a source of certificate authority data.
+type CABundleSource struct {
+  // Source of the certificate authority data, can be either "secret" or "configmap".
+  Kind string `json:"kind,omitempty" validate:"oneof=secret configmap"`
+  // Name of the secret or config map with the certificate authority data.
+  Name string `json:"name,omitempty"`
+  // Key in the named resource where the certificate authority data is present.
+  Key string `json:"key,omitempty"`
+}
+
+// Configuration for configuring TLS on various authenticators.
+type TLSSpec struct {
+	// X.509 Certificate Authority (base64-encoded PEM bundle). If omitted, a default set of system roots will be trusted.
+	// +optional
+	CertificateAuthorityData string `json:"certificateAuthorityData,omitempty"`
+	// +optional
+  CertificateAuthorityDataSource CABundleSource `json:"certificateAuthorityDataSource,omitempty"`
+}
+```
 
-* `WebhookAuthenticator.spec.tls.certificateAuthorityData`
-  * Code: `internal/controller/authenticator/webhookcachefiller/webhookcachefiller.go`
-  * Status Condition: `TLSConfigurationValid`
-* `JWTAuthenticator.spec.tls.certificateAuthorityData`
-  * Code: `internal/controller/authenticator/jwtcachefiller/jwtcachefiller.go`
-  * Status Condition: `TLSConfigurationValid`
+The `certificateAuthorityDataSource` field will be used only when no value is supplied in the `certificateAuthorityData` field. Supplying both is a configuration error.
 
-## Not applicable
+Users can use the new `certificateAuthorityDataSource` field to specify:
+1. Using the `kind` subfield, whether this is referencing a kubernetes secret or a configmap.
+   When using a secret, the secret must be of type `kubernetes.io/tls` or `Opaque`.
+2. Using the `name` subfield, the name of the kubernetes secret or configmap.
+   It is expected that this secret or configmap, if supplied, will exist in the same namespace where Pinniped is currently running.
+   Pinniped controllers do not seek access to secrets or configmap in a different namespace where sensitive information may be stored.
+3. Using the `key` subfield, the key within the secret or the configmap where the certificate authority data can be located.
```

Comment:
The `certificateAuthorityData` field's value is supposed to be base64 encoded. For the new `certificateAuthorityDataSource`, when the controller reads value at the given key from the ConfigMap or Secret, I think we are saying that we do not expect the data to be base64 encoded. So when the user creates the Secret using `stringData` or creates the ConfigMap, we do not want them to base64 encode the CA bundle, right? Because we think that these cert management tools will not base64 encode the value.

We should clarify that in this doc.

Reply:
The configmaps generated by tooling are not base64 encoded. However, the secrets generated by cert-manager are not base64 encoded and those generated by trust-manger are.
Reading secrets using the kubernetes client-go assures that the data read is decoded. I've added clarification that the config map data does not have to be base64 encoded.